proposals/gc/MVP.md

```diff
@@ -594,112 +594,54 @@ In particular, `ref.null` is typed as before, despite the introduction of `none`
+  - `ref.test null? ht : [(ref null ht')] -> [i32]`
+    - iff `ht <: ht'`
+  - if `null?` is present, returns 1 for null, otherwise 0
 
-* `ref.is_i31` checks whether a reference is an i31
-  - `ref.is_i31 : [anyref] -> [i32]`
+* `ref.cast null? <heaptype>` tries to convert to a given heap type
+  - `ref.cast null? ht : [(ref null ht')] -> [(ref null2? ht)]`
+    - iff `ht <: ht'`
+    - and `null? = null2?`
+  - traps if reference is not of requested type
+  - if `null?` is present, a null operand is passed through, otherwise traps on null
+  - equivalent to `(block $l (param anyref) (result (ref null? ht)) (br_on_cast null? ht $l) (unreachable))`
 
-* `br_on_data <labelidx>` branches if a reference is compound data
-  - `br_on_data $l : [t0* t] -> [t0* t]`
+* `br_on_cast <labelidx> null? <heaptype>` branches if a reference has a given heap type
+  - `br_on_cast $l null? ht : [t0* (ref null ht')] -> [t0* (ref null2? ht')]`
     - iff `$l : [t0* t']`
-    - and `t <: anyref`
-    - and `(ref data) <: t'`
-  - passes operand along with branch as data, plus possible extra args
-
-* `br_on_non_data <labelidx>` branches if a reference is not compound data
-  - `br_on_non_data $l : [t0* t] -> [t0* (ref data)]`
+    - and `(ref null3? ht) <: t'`
+    - and `ht <: tht` and `ht' <: tht` where `tht` is a common super type
+    - and `null? = null3? =/= null2?`
+  - passes operand along with branch under target type, plus possible extra args
+  - if `null?` is present, branches on null, otherwise does not
+
+* `br_on_cast_fail <labelidx> null? <heaptype>` branches if a reference does not have a given heap type
+  - `br_on_cast_fail $l null? ht : [t0* (ref null ht')] -> [t0* (ref null2? ht)]`
     - iff `$l : [t0* t']`
-    - and `t <: anyref`
-    - and `t <: t'`
+    - and `(ref null3? ht') <: t'`
+    - and `ht <: tht` and `ht' <: tht` where `tht` is a common super type
+    - and `null? = null2? =/= null3?`
   - passes operand along with branch, plus possible extra args
+  - if `null?` is present, does not branch on null, otherwise does
 
-* `br_on_array <labelidx>` branches if a reference is an array
-  - `br_on_array $l : [t0* t] -> [t0* t]`
-    - iff `$l : [t0* t']`
-    - and `t <: anyref`
-    - and `(ref array) <: t'`
-  - passes operand along with branch as data, plus possible extra args
+Note: Via subsumption to a common supertype, `ref.test` and `ref.cast` actually allow an operand of sibling reference type, even though this cannot possibly succeed. In the case of the branch instruction, this has to be made explicit, though, since the type also occurs in result position. The given typing rules allow this type to also propagate to the result, which avoids the need to compute a least upper bound between the operand type and the target type in the typing algorithm (allowing so is required to maintain subtype substitutability). In practice, it is sufficient to test that both types share the same top heap type.
```

Comment:
I don't understand the second or third sentences of this note. What has to be made explicit and what type are they referring to?

Reply:
@rossberg to be honest, it might be better to go back to the previous formulation of the side condition, even though `ref.test` and `ref.cast` can have a simpler formulation than `br_on_*`, just to avoid needing to explain that the conditions for all three instructions are "really" the same.

Reply:
Okay, I reverted the condition and simplified the note accordingly. Dropped the bit about the lub, since it is somewhat tangential and probably more confusing than helpful.